eZ/Bundle/EzPublishCoreBundle/EventListener/ViewControllerListener.php

```diff
@@ -72,77 +69,23 @@ public static function getSubscribedEvents()
-                $valueObject = $contentInfo;
-                $request->attributes->set('contentId', $valueObject->id);
-            }
-        } catch (UnauthorizedException $e) {
-            throw new AccessDeniedException();
-        }
-
-        if (!isset($valueObject)) {
-            $this->logger->error('Could not resolver a view controller, invalid value object to match.');
 
+        if (($viewBuilder = $this->viewBuilderRegistry->get($request->attributes->get('_controller'))) === null) {
             return;
         }
 
-        $requestParams = $request->attributes->get('params', []);
-        if ($valueObject instanceof Location && !isset($requestParams['location'])) {
-            $requestParams += ['location' => $valueObject];
-            $request->attributes->set('params', $requestParams);
-        }
-
-        $controllerReference = $this->controllerManager->getControllerReference(
-            $valueObject,
-            $request->attributes->get('viewType')
-        );
-
-        if ($valueObject instanceof Location && !$controllerReference instanceof ControllerReference) {
-            // If value object is a location and location view rules did not match a controller
-            // we should try matching with content view rules
-            $controllerReference = $this->controllerManager->getControllerReference(
-                $valueObject->contentInfo,
-                $request->attributes->get('viewType')
-            );
-        }
+        $parameterEvent = new ViewBuilderParameterCollectionEvent(clone $request);
+        $this->eventDispatcher->dispatch('view.builder_parameter_collection', $parameterEvent);
+        $view = $viewBuilder->buildView($parameterEvent->getParameters()->all());
 
-        if ($controllerReference instanceof ControllerReference) {
-            $request->attributes->set('_controller', $controllerReference->controller);
-            $event->setController($this->controllerResolver->getController($request));
+        $this->viewConfigurator->configure($view);
+        $request->attributes->set('view', $view);
 
-            return;
-        }
+        // View parameters are added as request attributes so that they are available as controller action parameters
+        $request->attributes->add($view->getParameters());
 
-        // if there is no custom controller, viewContent can be used instead of viewLocation.
-        if ($request->attributes->get('_controller') === 'ez_content:viewLocation') {
-            $request->attributes->set('_controller', 'ez_content:viewContent');
-            $event->setController($this->controllerResolver->getController($request));
-        }
-        if ($request->attributes->get('_controller') === 'ez_content:embedLocation') {
-            $request->attributes->set('_controller', 'ez_content:embedContent');
+        if ($view->getControllerReference() instanceof ControllerReference) {
```

Comment:
Nitpick: assign `$view->controllerReference()` to a variable that you can use below.